packages/overlay/src/OverlayDialog.ts

```diff
@@ -0,0 +1,207 @@
+/*
+Copyright 2020 Adobe. All rights reserved.
+This file is licensed to you under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License. You may obtain a copy
+of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software distributed under
+the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+OF ANY KIND, either express or implied. See the License for the specific language
+governing permissions and limitations under the License.
+*/
+import { ReactiveElement } from 'lit';
+import {
+    firstFocusableIn,
+    firstFocusableSlottedIn,
+} from '@spectrum-web-components/shared/src/first-focusable-in.js';
+import {
+    BeforetoggleClosedEvent,
+    BeforetoggleOpenEvent,
+    guaranteedAllTransitionend,
+    OpenableElement,
+    OverlayBase,
+} from './OverlayBase.js';
+import { VirtualTrigger } from './VirtualTrigger.js';
+import { OverlayOpenCloseDetail } from './overlay-types.js';
+
+type Constructor<T = Record<string, unknown>> = {
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    new (...args: any[]): T;
+    prototype: T;
+};
+
+function nextFrame(): Promise<void> {
+    return new Promise((res) => requestAnimationFrame(() => res()));
+}
+
+export function OverlayDialog<T extends Constructor<OverlayBase>>(
+    constructor: T
+): T & Constructor<ReactiveElement> {
+    class OverlayWithDialog extends constructor {
+        protected override async manageDialogOpen(): Promise<void> {
+            const targetOpenState = this.open;
+            await this.managePosition();
+            if (this.open !== targetOpenState) {
+                return;
+            }
+            await this.dialogEnsureOnDOM();
+            if (this.open !== targetOpenState) {
+                return;
+            }
+            const focusEl = await this.dialogMakeTransition(targetOpenState);
+            if (this.open !== targetOpenState) {
+                return;
+            }
+            await this.dialogApplyFocus(targetOpenState, focusEl);
+        }
+
+        protected async dialogEnsureOnDOM(): Promise<void> {
+            await nextFrame();
+            await nextFrame();
+            await nextFrame();
+            await nextFrame();
+        }
```

Comment:
Is there a more deterministic way to ensure it's on the DOM?

Reply:
I'll test some options...but it's quite weird x-browser, if I remember correctly.

Reply:
Unimportant for now then, just a nit.